backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java

```diff
@@ -1367,4 +1371,85 @@ public FutureProjectDAO(
             },
             executor);
   }
+
+  public InternalFuture<Void> setProjectShortName(SetProjectShortName request) {
+    // Request Parameter Validation
+    InternalFuture<Void> validateParamFuture =
+        InternalFuture.runAsync(
+            () -> {
+              if (request.getId().isEmpty() && request.getShortName().isEmpty()) {
+                throw new InvalidArgumentException(
+                    "Project ID and Project shortName not found in SetProjectShortName request");
+              } else if (request.getId().isEmpty()) {
+                throw new InvalidArgumentException(
+                    "Project ID not found in SetProjectShortName request");
+              } else if (request.getShortName().isEmpty()) {
+                throw new InvalidArgumentException(
+                    "Project shortName not found in SetProjectShortName request");
+              }
+            },
+            executor);
+
+    return validateParamFuture
+        .thenCompose(
+            unused ->
+                checkProjectPermission(
+                    request.getId(), ModelDBActionEnum.ModelDBServiceActions.UPDATE),
+            executor)
+        .thenCompose(
+            unused ->
+                getSelfAllowedResources(
+                    ModelDBResourceEnum.ModelDBServiceResourceTypes.PROJECT,
+                    ModelDBActionEnum.ModelDBServiceActions.READ,
+                    Collections.emptyList()),
+            executor)
+        .thenCompose(
+            selfAllowedResources -> {
+              InternalFuture<Optional<Long>> countFuture =
+                  InternalFuture.completedInternalFuture(Optional.of(0L));
+              if (!selfAllowedResources.isEmpty()) {
+                countFuture =
+                    jdbi.withHandle(
+                        handle ->
+                            handle
+                                .createQuery(
+                                    "select count(p.id) from project p where p.deleted = false AND p.short_name = :projectShortName AND p.id IN (<projectIds>)")
+                                .bind("projectShortName", request.getShortName())
+                                .bindList("projectIds", selfAllowedResources)
+                                .mapTo(Long.class)
+                                .findOne());
+              }
+              return countFuture.thenCompose(
+                  count -> {
+                    if (count.isPresent() && count.get() > 0) {
+                      return InternalFuture.failedStage(
+                          new AlreadyExistsException(
+                              "Project already exist with given short name"));
+                    }
+                    return InternalFuture.completedInternalFuture(null);
```

Comment:
I'm not sure this logic is correct. Let's say there's a workspace W and project names A and B. I only have access to A (e.g. I'm out of the org). Then if I try to rename the project to B, this check will succeed (I can't see that B exists), even though it's there.

Similarly, if I have projects named A in workspace W1 and B in workspace W2 and I have access to both, I should be able to rename B to A since they're in different workspaces.

Reply:
In the above both commented cases the above logic is works fine.

> I'm not sure this logic is correct. Let's say there's a workspace W and project names A and B. I only have access to A (e.g. I'm out of the org). Then if I try to rename the project to B, this check will succeed (I can't see that B exists), even though it's there.

we are checking read permission before count query like `checkProjectPermission(request.getId(), ModelDBActionEnum.ModelDBServiceActions.UPDATE),`  so if B is not accessible then it will throw Access Denined.

For second option:
We are fetching self allowed resources so we have both the workspace accessible ids and added it on above count query like: `.bindList("projectIds", selfAllowedResources)` 
```
.createQuery(
                                    "select count(p.id) from project p where p.deleted = false AND p.short_name = :projectShortName AND p.id IN (<projectIds>)")
                                .bind("projectShortName", request.getShortName())
                                .bindList("projectIds", selfAllowedResources)
                                .mapTo(Long.class)
                                .findOne());
```
so it will not allowed to rename B to A because both are in accessible ids.

Reply:
My first example was to rename `A` to `B`. Since I don't have access to `B`, B's ID won't be inside `projectIds` and `countFuture` will be zero. Even though there _is_ a project `B`, I just don't happen to have access to it.

Reply:
the old code had the same bug? : YES
Is there a test covering this that shows this behaviour and that it is unchanged? : Currently it is not but when we fix this i will add test for the same scenario.

Reply:
Created Jira ticket: https://vertaai.atlassian.net/browse/VUMM-202